File: src/main/java/szewek/flux/tile/FluxGenTile.java
package szewek.flux.tile;

import net.minecraft.block.BlockState;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.entity.player.PlayerInventory;
import net.minecraft.fluid.Fluid;
import net.minecraft.inventory.ItemStackHelper;
import net.minecraft.inventory.container.Container;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.CompoundNBT;
import net.minecraft.tileentity.ITickableTileEntity;
import net.minecraft.tileentity.LockableTileEntity;
import net.minecraft.util.Direction;
import net.minecraft.util.IIntArray;
import net.minecraft.util.NonNullList;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.text.ITextComponent;
import net.minecraft.util.text.TranslationTextComponent;
import net.minecraft.world.Explosion;
import net.minecraft.world.World;
import net.minecraftforge.common.ForgeHooks;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.util.LazyOptional;
import net.minecraftforge.common.util.NonNullSupplier;
import net.minecraftforge.energy.CapabilityEnergy;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.fluids.capability.CapabilityFluidHandler;
import net.minecraftforge.fluids.capability.IFluidHandler;
import net.minecraftforge.registries.ForgeRegistries;
import net.minecraftforge.registries.ForgeRegistry;
import szewek.fl.util.FluidsUtil;
import szewek.fl.util.IntPair;
import szewek.flux.F;
import szewek.flux.FluxCfg;
import szewek.flux.container.FluxGenContainer;
import szewek.flux.data.FluxGenValues;
import szewek.flux.energy.EnergyCache;
import szewek.flux.tile.part.GeneratorEnergy;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;

public class FluxGenTile extends LockableTileEntity implements ITickableTileEntity {
	public static final int fluidCap = 4000;
	private final EnergyCache energyCache = new EnergyCache(this);
	private final NonNullList<ItemStack> items = NonNullList.withSize(2, ItemStack.EMPTY);
	private final AtomicBoolean isDirty = new AtomicBoolean();
	private final Tank tank = new Tank(isDirty);
	private final GeneratorEnergy energy = new GeneratorEnergy(1_000_000);
	private final WorkValues work = new WorkValues(isDirty);
	private int redstoneState = -1;
	protected final IIntArray tileData = new IIntArray() {
		@Override
		public int get(int i) {
			if (i < 2) return energy.getEnergy16Bit(i == 1);
			if (i < 6) return tank.getData(i - 2);
			return work.getData(i - 6);
		}

		@Override
		public void set(int i, int v) {
			if (i < 2) energy.setEnergy16Bit(i == 1, v);
			else if (i < 6) tank.setData(i - 2, v);
			else work.setData(i - 6, v);
		}

		@Override
		public int getCount() {
			return 10;
		}
	};

	public FluxGenTile() {
		super(F.T.FLUXGEN);
	}

	@Override
	public void load(BlockState blockState, CompoundNBT compound) {
		super.load(blockState, compound);
		energy.readNBT(compound);
		work.readNBT(compound);
		ItemStackHelper.loadAllItems(compound, items);
		List<FluidStack> fluidList = NonNullList.withSize(tank.fluids.length, FluidStack.EMPTY);
		FluidsUtil.loadAllFluids(compound, fluidList);
		for (int i = 0; i < tank.fluids.length; i++) {
			tank.fluids[i] = fluidList.get(i);
		}
	}

	@Override
	public CompoundNBT save(CompoundNBT compound) {
		super.save(compound);
		energy.writeNBT(compound);
		work.writeNBT(compound);
		ItemStackHelper.saveAllItems(compound, items);
		FluidsUtil.saveAllFluids(compound, Arrays.asList(tank.fluids), true);

		return compound;
	}

	public static void tick(World w, BlockPos bp, BlockState state, FluxGenTile it) {
		if (w.isClientSide) return;
		if (it.redstoneState == -1) {
			it.redstoneState = w.getBestNeighborSignal(bp) > 0 ? 1 : 0;
		}
		final GeneratorEnergy en = it.energy;
		if (it.redstoneState == 0) {
			final WorkValues wv = it.work;
			if (wv.canBegin(ForgeHooks.getBurnTime(it.items.get(0)))) {
				wv.update(it.items, it.tank.fluids);
			} else if (en.generate(it.work.gen)) {
				wv.tick();
				it.setChanged();
			} else if (wv.gen > en.getMaxEnergyStored()) {
				// BLOW UP
				w.explode(null, bp.getX() + 0.5, bp.getY() + 0.5, bp.getZ() + 0.5, 9.0F, Explosion.Mode.DESTROY);
			}
		}
		en.share(it.energyCache);
		if (it.isDirty.getAndSet(false)) it.setChanged();
	}

	@Override
	public void tick() {
		tick(level, worldPosition, /* unused */ null, this);
	}

	public void setRedstoneState(boolean state) {
		redstoneState = state ? 1 : 0;
	}

	public Tank getTank() {
		return tank;
	}

	@Override
	public <T> LazyOptional<T> getCapability(Capability<T> cap, Direction dir) {
		if (!remove) {
			if (cap == CapabilityEnergy.ENERGY) {
				return energy.lazyCast();
			} else if (cap == CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY) {
				return tank.lazy.cast();
			}
		}
		return super.getCapability(cap, dir);
	}

	@Override
	public void setRemoved() {
		super.setRemoved();
		energyCache.clear();
		energy.invalidate();
		tank.lazy.invalidate();
	}

	@Override
	public int getContainerSize() {
		return 2;
	}

	@Override
	public boolean isEmpty() {
		return items.get(0).isEmpty() && items.get(1).isEmpty();
	}

	@Override
	public ItemStack getItem(int i) {
		if (i < 0 || i >= items.size())
			throw new IndexOutOfBoundsException("Getting slot " + i + " outside range [0," + items.size() + ")");
		return items.get(i);
	}

	@Override
	public ItemStack removeItem(int i, int count) {
		return i >= 0 && i <= items.size() && count > 0 && !items.get(i).isEmpty() ? items.get(i).split(count) : ItemStack.EMPTY;
	}

	@Override
	public ItemStack removeItemNoUpdate(int i) {
		if (i >= 0 && i <= items.size()) {
			ItemStack stack = items.get(i);
			items.set(i, ItemStack.EMPTY);
			return stack;
		}
		return ItemStack.EMPTY;
	}

	@Override
	public void setItem(int i, ItemStack stack) {
		if (i >= 0 && i <= items.size()) {
			if (stack.getCount() > 64) stack.setCount(64);
			items.set(i, stack);
		}
	}

	@Override
	public boolean stillValid(PlayerEntity player) {
		return player.level.getBlockEntity(worldPosition) == this && worldPosition.distSqr(player.position(), true) <= 64.0;
	}

	@Override
	public void clearContent() {
		items.clear();
	}

	@Override
	protected ITextComponent getDefaultName() {
		return new TranslationTextComponent("container.flux.fluxgen");
	}

	@Override
	protected Container createMenu(int id, PlayerInventory playerInv) {
		return new FluxGenContainer(id, playerInv, this, tileData);
	}

	static class Tank implements IFluidHandler, NonNullSupplier<IFluidHandler> {
		private final FluidStack[] fluids = {FluidStack.EMPTY, FluidStack.EMPTY};
		private final LazyOptional<IFluidHandler> lazy = LazyOptional.of(this);
		private final AtomicBoolean isDirty;

		Tank(AtomicBoolean dirty) {
			isDirty = dirty;
		}

		private int getData(int i) {
			if (i >= 4) return 0;
			FluidStack fs = fluids[i >> 1];
			if (i % 2 == 0) {
				return ((ForgeRegistry<Fluid>) ForgeRegistries.FLUIDS).getID(fs.getFluid());
			}
			return fs.getAmount();
		}
		private void setData(int i, int v) {
			if (i >= 4) return;
			FluidStack fs = fluids[i >> 1];
			if (i % 2 == 0) {
				fluids[i >> 1] = new FluidStack(((ForgeRegistry<Fluid>) ForgeRegistries.FLUIDS).getValue(v), fs.getAmount());
			} else {
				if (!fs.isEmpty()) {
					fs.setAmount(v);
				}
			}
		}

		@Override
		public int getTanks() {
			return 2;
		}

		@Override
		public FluidStack getFluidInTank(int tank) {
			return fluids[tank];
		}

		@Override
		public int getTankCapacity(int tank) {
			return fluidCap;
		}

		@Override
		public boolean isFluidValid(int tank, FluidStack stack) {
			return false;
		}

		@Override
		public int fill(FluidStack resource, FluidAction action) {
			if (resource.getAmount() <= 0) {
				return 0;
			}
			int s;
			if (FluxGenValues.HOT_FLUIDS.has(resource.getFluid())) {
				s = 0;
			} else if (FluxGenValues.COLD_FLUIDS.has(resource.getFluid())) {
				s = 1;
			} else {
				return 0;
			}
			FluidStack fs = fluids[s];
			if (!fs.isEmpty() && !fs.isFluidEqual(resource)) {
				return 0;
			}
			int l = fluidCap - fs.getAmount();
			if (l > resource.getAmount()) {
				l = resource.getAmount();
			}
			if (l > 0 && action.execute()) {
				if (fs.isEmpty()) {
					fluids[s] = resource.copy();
				} else {
					fs.grow(l);
				}
				isDirty.set(true);
			}
			return l;
		}

		@Override
		public FluidStack drain(FluidStack resource, FluidAction action) {
			return FluidStack.EMPTY;
		}

		@Override
		public FluidStack drain(int maxDrain, FluidAction action) {
			return FluidStack.EMPTY;
		}

		@Override
		public IFluidHandler get() {
			return this;
		}
	}

	static class WorkValues {
		private final AtomicBoolean isDirty;
		private int ticks, max, gen, speed;

		WorkValues(AtomicBoolean isDirty) {
			this.isDirty = isDirty;
		}

		private int getData(int i) {
			switch (i) {
				case 0: return ticks;
				case 1: return max;
				case 2: return gen;
				case 3: return speed;
			}
			return 0;
		}

		private void setData(int i, int v) {
			switch (i) {
				case 0: ticks = v; break;
				case 1: max = v; break;
				case 2: gen = v; break;
				case 3: speed = v; break;
			}
		}

		private void readNBT(CompoundNBT nbt) {
			ticks = nbt.getInt("WorkTicks");
			max = nbt.getInt("MaxWork");
			gen = nbt.getInt("Gen");
			speed = nbt.getInt("WorkSpeed");
		}

		private void writeNBT(CompoundNBT nbt) {
			nbt.putInt("WorkTicks", ticks);
			nbt.putInt("MaxWork", max);
			nbt.putInt("Gen", gen);
			nbt.putInt("WorkSpeed", speed);
		}

		private boolean canBegin(int burnValue) {
			return (max == 0 && burnValue > 0) || ticks >= max;
		}

		private void tick() {
			ticks += speed;
			if (max <= ticks) {
				max = 0;
				gen = 0;
			}
		}

		private void update(NonNullList<ItemStack> items, FluidStack[] fluids) {
			ticks = 0;
			ItemStack fuel = items.get(0);
			int f = ForgeHooks.getBurnTime(fuel);
			if (f == 0) {
				max = 0;
				return;
			}
			ItemStack catalyst = items.get(1);
			gen = FluxCfg.ENERGY.fluxGenBaseEnergy.get();
			IntPair genCat = FluxGenValues.CATALYSTS.get(catalyst.getItem());
			if (genCat.r <= catalyst.getCount()) {
				gen *= genCat.l;
				if (genCat.r > 0) catalyst.grow(-genCat.r);
			}
			IntPair genHot = FluxGenValues.HOT_FLUIDS.get(fluids[0].getFluid());
			if (genHot.r <= fluids[0].getAmount()) {
				f *= genHot.l;
				if (genHot.r > 0) fluids[0].grow(-genHot.r);
			}
			IntPair genCold = FluxGenValues.COLD_FLUIDS.get(fluids[1].getFluid());
			if (genCold.r <= fluids[1].getAmount()) {
				speed = genCold.l < genCat.l ? genCat.l - genCold.l : 1;
				gen *= speed;
				if (genCold.r > 0) fluids[1].grow(-genCold.r);
			} else {
				speed = 1;
			}
			fuel.grow(-1);
			max = f;
			isDirty.set(true);
		}
	}
}

File: src/main/java/szewek/flux/tile/FluxGenTile.kt
package szewek.flux.tile

import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.inventory.IInventory
import net.minecraft.inventory.ItemStackHelper
import net.minecraft.inventory.container.Container
import net.minecraft.inventory.container.INamedContainerProvider
import net.minecraft.item.ItemStack
import net.minecraft.nbt.CompoundNBT
import net.minecraft.tileentity.ITickableTileEntity
import net.minecraft.tileentity.LockableTileEntity
import net.minecraft.util.Direction
import net.minecraft.util.IIntArray
import net.minecraft.util.NonNullList
import net.minecraft.util.text.ITextComponent
import net.minecraft.util.text.TranslationTextComponent
import net.minecraftforge.common.ForgeHooks
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.util.LazyOptional
import net.minecraftforge.energy.CapabilityEnergy
import net.minecraftforge.energy.IEnergyStorage
import net.minecraftforge.fluids.FluidStack
import net.minecraftforge.fluids.capability.CapabilityFluidHandler
import net.minecraftforge.fluids.capability.IFluidHandler
import net.minecraftforge.fluids.capability.IFluidHandler.FluidAction
import net.minecraftforge.items.IItemHandler
import net.minecraftforge.items.ItemHandlerHelper
import szewek.flux.FTiles
import szewek.flux.container.FluxGenContainer
import szewek.flux.recipe.FluxGenRecipes.getCatalyst
import szewek.flux.recipe.FluxGenRecipes.getColdFluid
import szewek.flux.recipe.FluxGenRecipes.getHotFluid
import szewek.flux.recipe.FluxGenRecipes.isCatalyst
import szewek.flux.recipe.FluxGenRecipes.isColdFluid
import szewek.flux.recipe.FluxGenRecipes.isHotFluid

class FluxGenTile : LockableTileEntity(FTiles.FLUXGEN), IInventory, IItemHandler, IFluidHandler, ITickableTileEntity, INamedContainerProvider, IEnergyStorage {
    private val items = NonNullList.withSize(2, ItemStack.EMPTY)
    private val fluids = arrayOf(FluidStack.EMPTY, FluidStack.EMPTY)
    private var tickCount = 0
    private var energy = 0
    private var workTicks = 0
    private var maxWork = 0
    private var energyGen = 0
    private var workSpeed = 0
    private var isReady = false
    private var isDirty = false
    @JvmField
    var receivedRedstone = false
    private val fluxGenData: IIntArray = object : IIntArray {
        override fun get(i: Int): Int = when (i) {
            0 -> energy
            1 -> workTicks
            2 -> maxWork
            3 -> energyGen
            4 -> workSpeed
            else -> 0
        }

        override fun set(i: Int, v: Int) {
            when (i) {
                0 -> {
                    energy = v
                    workTicks = v
                    maxWork = v
                    energyGen = v
                    workSpeed = v
                }
                1 -> {
                    workTicks = v
                    maxWork = v
                    energyGen = v
                    workSpeed = v
                }
                2 -> {
                    maxWork = v
                    energyGen = v
                    workSpeed = v
                }
                3 -> {
                    energyGen = v
                    workSpeed = v
                }
                4 -> workSpeed = v
            }
        }

        override fun size() = 5
    }
    private val selfHandler = LazyOptional.of { this }
    override fun read(compound: CompoundNBT) {
        super.read(compound)
        energy = compound.getInt("E")
        if (energy >= maxEnergy) energy = maxEnergy
        workTicks = compound.getInt("WorkTicks")
        maxWork = compound.getInt("MaxWork")
        energyGen = compound.getInt("Gen")
        workSpeed = compound.getInt("WorkSpeed")
        ItemStackHelper.loadAllItems(compound, items)
    }

    override fun write(compound: CompoundNBT): CompoundNBT {
        super.write(compound)
        compound.putInt("E", energy)
        compound.putInt("WorkTicks", workTicks)
        compound.putInt("MaxWork", maxWork)
        compound.putInt("Gen", energyGen)
        compound.putInt("WorkSpeed", workSpeed)
        ItemStackHelper.saveAllItems(compound, items)
        return compound
    }

    override fun tick() {
        if (world == null || world!!.isRemote) return
        if (!isReady) {
            if (world!!.getRedstonePowerFromNeighbors(pos) > 0) receivedRedstone = true
            isReady = true
        }
        if (!receivedRedstone) {
            if (maxWork == 0 && ForgeHooks.getBurnTime(items[0]) > 0 || workTicks >= maxWork) {
                workTicks = 0
                maxWork = updateWork()
            } else if (energy + energyGen <= maxEnergy) {
                energy += energyGen
                workTicks += workSpeed
                if (maxWork <= workTicks) {
                    maxWork = 0
                    energyGen = 0
                }
            }
        }
        tickCount++
        if (tickCount > 3 && energy > 0) {
            tickCount = 0
            for (d in Direction.values()) {
                val bp = pos.offset(d, 1)
                val te = world!!.getTileEntity(bp)
                te?.getCapability(CapabilityEnergy.ENERGY, d.opposite)?.ifPresent { ie: IEnergyStorage -> sendEnergyTo(ie) }
            }
        }
        if (isDirty) markDirty()
    }

    private fun updateWork(): Int {
        val fuel = items[0]
        var f = ForgeHooks.getBurnTime(fuel)
        if (f == 0) return 0
        val catalyst = items[1]
        val genCat = getCatalyst(catalyst.item)
        val genHot = getHotFluid(fluids[0])
        val genCold = getColdFluid(fluids[1])
        energyGen = 40
        if (genCat.usage <= catalyst.count) {
            energyGen *= genCat.factor
            catalyst.grow(-genCat.usage)
        }
        if (genHot.usage <= fluids[0].amount) {
            f *= genHot.factor
            fluids[0].grow(-genHot.usage)
        }
        if (genCold.usage <= fluids[1].amount) {
            workSpeed = if (genCold.factor < genCat.factor) genCold.factor - genCat.factor else 1
            fluids[1].grow(-genCold.usage)
        } else {
            workSpeed = 1
        }
        fuel.grow(-1)
        isDirty = true
        return f
    }

    override fun <T> getCapability(cap: Capability<T>, dir: Direction?): LazyOptional<T>? {
        if (!removed) {
            if (cap === CapabilityEnergy.ENERGY || cap === CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY) return selfHandler.cast()
        }
        return super.getCapability(cap, dir)
    }

    override fun remove() {
        super.remove()
        selfHandler.invalidate()
    }

    override fun getEnergyStored() = energy
    override fun getMaxEnergyStored() = maxEnergy

    override fun canExtract() = true
    override fun canReceive() = false

    override fun receiveEnergy(maxReceive: Int, simulate: Boolean): Int {
        return 0
    }

    override fun extractEnergy(maxExtract: Int, simulate: Boolean): Int {
        var r = maxExtract
        if (r > energy) r = energy
        if (!simulate) {
            energy -= r
            isDirty = true
        }
        return r
    }

    private fun sendEnergyTo(ie: IEnergyStorage) {
        if (ie.canReceive()) {
            var r = sendAmount
            if (r >= energy) r = energy
            r = ie.receiveEnergy(r, true)
            if (r > 0) {
                energy -= r
                isDirty = true
                ie.receiveEnergy(r, false)
            }
        }
    }

    override fun getSlots(): Int {
        return 2
    }

    override fun getSizeInventory(): Int {
        return 2
    }

    override fun isEmpty(): Boolean {
        return items[0].isEmpty && items[1].isEmpty
    }

    override fun getStackInSlot(i: Int): ItemStack {
        if (i < 0 || i >= items.size) throw RuntimeException("Getting slot " + i + " outside range [0," + items.size + ")")
        return items[i]
    }

    override fun decrStackSize(i: Int, count: Int): ItemStack {
        return if (i >= 0 && i <= items.size && count > 0 && !items[i].isEmpty) items[i].split(count) else ItemStack.EMPTY
    }

    override fun removeStackFromSlot(i: Int): ItemStack {
        if (i >= 0 && i <= items.size) {
            val stack = items[i]
            items[i] = ItemStack.EMPTY
            return stack
        }
        return ItemStack.EMPTY
    }

    override fun setInventorySlotContents(i: Int, stack: ItemStack) {
        if (i >= 0 && i <= items.size) {
            if (stack.count > 64) stack.count = 64
            items[i] = stack
        }
    }

    override fun getInventoryStackLimit(): Int {
        return 64
    }

    override fun isUsableByPlayer(player: PlayerEntity): Boolean {
        return player.world.getTileEntity(pos) === this && pos.distanceSq(player.positionVec, true) <= 64.0
    }

    override fun clear() {
        items.clear()
    }

    override fun insertItem(slot: Int, stack: ItemStack, simulate: Boolean): ItemStack {
        if (slot < 0 || slot >= items.size) throw RuntimeException("Getting slot " + slot + " outside range [0," + items.size + ")")
        if (stack.isEmpty) return ItemStack.EMPTY
        if (slot == 0 && ForgeHooks.getBurnTime(stack) == 0 || slot == 1 && !isCatalyst(stack.item)) {
            return stack
        }
        var l = stack.maxStackSize
        if (l > 64) l = 64
        val sc = stack.count
        val xis = items[slot]
        if (!xis.isEmpty) {
            if (!ItemHandlerHelper.canItemStacksStack(stack, xis)) return stack
            l -= xis.count
        }
        if (0 >= l) return stack
        val rl = sc > l
        if (!simulate) {
            if (xis.isEmpty) items[slot] = if (rl) ItemHandlerHelper.copyStackWithSize(stack, l) else stack else xis.grow(if (rl) l else sc)
            isDirty = true
        }
        return if (rl) ItemHandlerHelper.copyStackWithSize(stack, sc - l) else ItemStack.EMPTY
    }

    override fun extractItem(slot: Int, amount: Int, simulate: Boolean): ItemStack {
        return ItemStack.EMPTY
    }

    override fun getSlotLimit(slot: Int) = 64

    override fun isItemValid(slot: Int, stack: ItemStack): Boolean {
        return false
    }

    override fun getTanks() = 2

    override fun getFluidInTank(tank: Int): FluidStack = fluids[tank]

    override fun getTankCapacity(tank: Int) = fluidCap

    override fun isFluidValid(tank: Int, stack: FluidStack) = false

    override fun fill(resource: FluidStack, action: FluidAction): Int {
        if (resource.amount <= 0) return 0
        var s = -1
        if (isHotFluid(resource)) s = 0 else if (isColdFluid(resource)) s = 1
        if (s == -1 || !fluids[s].isFluidEqual(resource)) return 0
        var l = fluidCap - fluids[s].amount
        if (l > resource.amount) l = resource.amount
        if (l > 0 && action.execute()) {
            if (fluids[s].isEmpty) fluids[s] = resource.copy() else fluids[s].grow(l)
            isDirty = true
        }
        return l
    }

    override fun drain(resource: FluidStack, action: FluidAction): FluidStack {
        return FluidStack.EMPTY
    }

    override fun drain(maxDrain: Int, action: FluidAction): FluidStack {
        return FluidStack.EMPTY
    }

    override fun getDefaultName(): ITextComponent {
        return TranslationTextComponent("container.flux.fluxgen")
    }

    override fun createMenu(id: Int, playerInv: PlayerInventory): Container {
        return FluxGenContainer(id, playerInv, this, fluxGenData)
    }

    companion object {
        const val maxEnergy = 1000000
        const val sendAmount = 40000
        const val fluidCap = 4000
    }
}